File: internal/cli/validate.go
package cli

import (
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"github.com/daFish/gogo-meta/internal/config"
	"github.com/daFish/gogo-meta/internal/output"
	"github.com/spf13/cobra"
)

func newValidateCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "validate",
		Short: "Validate config files and check that configured projects exist in the working copy",
		RunE:  runValidate,
	}
}

type validationResult struct {
	file  string
	valid bool
	err   string
}

func runValidate(_ *cobra.Command, _ []string) error {
	cwd, err := os.Getwd()
	if err != nil {
		return err
	}

	configFiles, err := findConfigFiles(cwd)
	if err != nil {
		return err
	}

	if len(configFiles) == 0 {
		output.Warning("No config files found in current directory")
		return nil
	}

	var results []validationResult
	for _, filename := range configFiles {
		filePath := filepath.Join(cwd, filename)
		results = append(results, validateConfigFile(filePath, filename))
	}

	configHasErrors := false
	for _, r := range results {
		if r.valid {
			output.ProjectStatus(r.file, "success", "")
		} else {
			output.ProjectStatus(r.file, "error", r.err)
			configHasErrors = true
		}
	}

	workingCopyHasErrors := validateWorkingCopy(cwd)

	if configHasErrors || workingCopyHasErrors {
		return fmt.Errorf("validation failed")
	}
	return nil
}

func findConfigFiles(cwd string) ([]string, error) {
	entries, err := os.ReadDir(cwd)
	if err != nil {
		return nil, err
	}

	var configFiles []string
	for _, entry := range entries {
		name := entry.Name()
		if name == ".gogo" || strings.HasPrefix(name, ".gogo.") {
			configFiles = append(configFiles, name)
		}
	}

	sort.Strings(configFiles)
	return configFiles, nil
}

const missingDirectoryHint = "directory missing — run 'gogo migrate' if it moved, or 'gogo git update' to clone"

// validateWorkingCopy reports whether any configured project directory is
// missing from the working copy. It prints per-project errors and returns true
// when at least one directory is missing. If the cwd is not inside a meta repo
// (or there are no projects), it returns false without output.
func validateWorkingCopy(cwd string) bool {
	result, err := config.ReadMetaConfig(cwd, nil)
	if err != nil {
		return false
	}

	projectPaths := make([]string, 0, len(result.Config.Projects))
	for p := range result.Config.Projects {
		projectPaths = append(projectPaths, p)
	}
	if len(projectPaths) == 0 {
		return false
	}
	sort.Strings(projectPaths)

	hasErrors := false
	for _, projectPath := range projectPaths {
		projectDir := filepath.Join(result.MetaDir, projectPath)
		if !config.FileExists(projectDir) {
			output.ProjectStatus(projectPath, "error", missingDirectoryHint)
			hasErrors = true
		}
	}

	if !hasErrors {
		output.Success(fmt.Sprintf("All %d project directories present", len(projectPaths)))
	}

	return hasErrors
}

func validateConfigFile(filePath, filename string) validationResult {
	format := config.DetectFormat(filePath)

	content, err := os.ReadFile(filePath)
	if err != nil {
		return validationResult{file: filename, valid: false, err: err.Error()}
	}

	cfg, err := config.ParseConfigContent(content, format)
	if err != nil {
		return validationResult{file: filename, valid: false, err: fmt.Sprintf("Invalid %s: %v", format, err)}
	}

	if err := config.Validate(*cfg); err != nil {
		return validationResult{file: filename, valid: false, err: fmt.Sprintf("Invalid structure: %v", err)}
	}

	return validationResult{file: filename, valid: true}
}

File: src/commands/validate.ts
import { Command } from 'commander';
import { readFile, readdir } from 'node:fs/promises';
import { join } from 'node:path';
import { parse as parseYaml } from 'yaml';
import { MetaConfigSchema, type MetaConfig } from '../types/index.js';
import { detectFormat, readMetaConfig, fileExists } from '../core/config.js';
import * as output from '../core/output.js';

interface ValidationResult {
  file: string;
  valid: boolean;
  error?: string;
}

const MISSING_DIRECTORY_HINT =
  "directory missing — run 'gogo migrate' if it moved, or 'gogo git update' to clone";

function isGogoConfigFile(filename: string): boolean {
  return filename === '.gogo' || filename.startsWith('.gogo.');
}

export async function findConfigFiles(cwd: string): Promise<string[]> {
  const entries = await readdir(cwd);
  const configFiles: string[] = [];

  for (const entry of entries) {
    if (isGogoConfigFile(entry)) {
      configFiles.push(entry);
    }
  }

  return configFiles.sort();
}

export async function validateCommand(): Promise<void> {
  const cwd = process.cwd();
  const results: ValidationResult[] = [];

  const configFiles = await findConfigFiles(cwd);

  for (const filename of configFiles) {
    const filePath = join(cwd, filename);
    results.push(await validateConfigFile(filePath, filename));
  }

  if (results.length === 0) {
    output.warning('No config files found in current directory');
    return;
  }

  const hasErrors = results.some((r) => !r.valid);

  for (const result of results) {
    if (result.valid) {
      output.projectStatus(result.file, 'success');
    } else {
      output.projectStatus(result.file, 'error', result.error);
    }
  }

  const workingCopyHasErrors = await validateWorkingCopy(cwd);

  if (hasErrors || workingCopyHasErrors) {
    throw new Error('Validation failed');
  }
}

async function validateWorkingCopy(cwd: string): Promise<boolean> {
  let config: MetaConfig;
  let metaDir: string;
  try {
    ({ config, metaDir } = await readMetaConfig(cwd));
  } catch {
    return false;
  }

  const projectPaths = Object.keys(config.projects);
  if (projectPaths.length === 0) {
    return false;
  }

  let hasErrors = false;
  for (const projectPath of projectPaths) {
    const projectDir = join(metaDir, projectPath);
    if (!(await fileExists(projectDir))) {
      output.projectStatus(projectPath, 'error', MISSING_DIRECTORY_HINT);
      hasErrors = true;
    }
  }

  if (!hasErrors) {
    output.success(`All ${projectPaths.length} project directories present`);
  }

  return hasErrors;
}

async function validateConfigFile(filePath: string, filename: string): Promise<ValidationResult> {
  const format = detectFormat(filePath);

  try {
    const content = await readFile(filePath, 'utf-8');
    const parsed = format === 'yaml' ? parseYaml(content) : JSON.parse(content);
    MetaConfigSchema.parse(parsed);
    return { file: filename, valid: true };
  } catch (error) {
    if (error instanceof SyntaxError) {
      return { file: filename, valid: false, error: 'Invalid JSON' };
    }
    if (error instanceof Error && error.name === 'YAMLParseError') {
      return { file: filename, valid: false, error: 'Invalid YAML' };
    }
    if (error instanceof Error && error.name === 'ZodError') {
      return { file: filename, valid: false, error: `Invalid structure: ${error.message}` };
    }
    return { file: filename, valid: false, error: String(error) };
  }
}

export function registerValidateCommand(program: Command): void {
  program
    .command('validate')
    .description('Validate config files and check that configured projects exist in the working copy')
    .action(async () => {
      await validateCommand();
    });
}
